api/REST-API.md

```diff
@@ -1,191 +1,600 @@
+
+So, if we continue with our base url example above, these example resources could be reached as such:
+
+|Resource|Noun|Base URI|RESTful route|
+|-|-|-|-|
+|Users of the application|`users`|`/api/v1/`|`/api/v1/users`|
+|Products|`products`|`/api/v1/`|`/api/v1/products`|
+|Customers accounts|`accounts`|`/api/v1/`|`/api/v1/accounts`|
+
+When building a more advanced API, with a database that has relations between records, you will still keep the same format, and use a forward slash `/` to separate hierarchical relationships.
+
+For example, with `customers` having a relationship in a database with the `invoices` resource, and these `invoices` have in their turn a relationship with `products`, you could see something like this:
+
+```
+/api/v1/customers
+/api/v1/customers/invoices
+/api/v1/customers/invoices/{id}
+/api/v1/customers/invoices/{id}/products
+/api/v1/customers/invoices/{id}/products/{id}
 ```
 
-Or consider the operation `addEvent`.
+#### PUT vs POST
 
-##### `addEvent`
+Sometimes developers experience some difficulty in understanding the difference between PUT and POST. Although there is not really an answer carved in stone, [this nice article](https://restfulapi.net/rest-put-vs-post/) has a neat chart laying out the specific differences between each one, so we retranscribe here for your convenience:
+
+|`PUT`|`POST`|
+|--|--|
+| RFC-2616 clearly mention that `PUT` method requests for the enclosed entity be stored under the supplied Request-URI. If the Request-URI refers to an already existing resource – an update operation will happen, otherwise create operation should happen if Request-URI is a valid resource URI (assuming client is allowed to determine resource identifier). `PUT` /questions/{question-id} | The `POST` method is used to request that the origin server accept the entity enclosed in the request as a new subordinate of the resource identified by the Request-URI in the Request-Line. It essentially means that `POST` request-URI should be of a collection URI. `POST` /questions |
+| `PUT` method is idempotent. So if you send retry a request multiple times, that should be equivalent to single request modification.| `POST` is NOT idempotent. So if you retry the request N times, you will end up having N resources with N different URIs created on server.|
+| Use `PUT` also when you want to modify a singular resource which is already a part of resources collection. |Use `POST` when you want to add a child resource under resources collection.|
+| Though `PUT` is idempotent, we shall not cache it’s response.| Responses to this method are not cacheable, unless the response includes appropriate Cache-Control or Expires header fields. However, the 303 (See Other) response can be used to direct the user agent to retrieve a cacheable resource.|
+| Generally, in practice, always use `PUT` for `UPDATE` operations.| Always use `POST` for `CREATE` operations.|
+
+*Ref: (https://restfulapi.net/rest-put-vs-post/), accessed on 2020-11-05.*
```

Comment:
Linking to this definition is fine but pasting it here actually just adds unnecessary detail. The paragraph that follows in your own words is far better so I would just remove the inlining here.

Reply:
Absolutely.

Removed in [54ad163](https://github.com/Techtonica/curriculum/pull/1410/commits/54ad163de6ebb01deeccbe0246df2d01054aa7fb).